### DISCUSSION TITLE
BM25 Python Client

### DISCUSSION THREAD

**Discussion by raphaelsty** [duckdb/duckdb#13898] [Show and tell]
Hi,

Congratulation for the good workd, DuckDB is amazing!!

I did release a small tool in Python called ducksearch which is based on DuckDB fts extension. I made few updates to it to pre-compute BM25 scores so it's faster. I also used pyarrow to make it batch-compatible and introduced an approximation to get more QPS in batch

https://github.com/lightonai/ducksearch

The search would benefit from a TOP N GROUP BY function which would not compute the exact order but rather select the top n with O(n) complexity.

Thank you :) 

Raphael

**Comment by Maxxen** [comment]
In DuckDB v1.1.0 The `min_by` and `max_by` aggregate functions take an optional third `n` argument now which makes TOP N GROUP BY a lot easier :)

**Reply by raphaelsty** [reply]
Amazing ⚡️
